### PR TITLE
[Cart] Cart with multiple tax rates scenarios

### PR DESCRIPTION
**Pull request by Zales0123**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets |
| License       | MIT
| Doc PR        |

Based on https://github.com/Sylius/Sylius/pull/3993

### CODE REVIEW

File: features/shop/cart_with_single_tax_rate.feature [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        And my cart taxes should be "€23.00"
+
+    Scenario: Proper taxes for multiple same products with the same tax rate
+        When I add 3 products "PHP T-Shirt" to the cart
+        Then my cart total should be "€369.00"
+        And my cart taxes should be "€69.00"
+
+    Scenario: Proper taxes for multiple different products with the same tax rate
+        When I add 3 products "PHP T-Shirt" to the cart
+        When I add 2 products "Symfony Hat" to the cart
```
Comment (stloyd) [inline features/shop/cart_with_single_tax_rate.feature]: `And`

File: src/Sylius/Behat/Context/Setup/UserContext.php [hunk 2]
```diff
@@ -65,19 +76,108 @@ public function __construct(
...
+    }
+
+    /**
+     * @param CustomerInterface $customer
+     * @param string $email
+     * @param string $password
+     *
+     * @return UserInterface
+     */
+    private function createUser(CustomerInterface $customer, $email = 'john.doe@example.com', $password = 'testPassword')
```
Comment (stloyd) [inline src/Sylius/Behat/Context/Setup/UserContext.php]: Please be consistent ;) Once you use multi-line (see below) once you put it all in one line :)

File: src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactoryInterface.php [hunk 3]
```diff
@@ -25,4 +25,11 @@
      * @return ZoneInterface
      */
     public function createTyped($type);
+
+    /**
+     * @param array $members
```
Comment (stloyd) [inline src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactoryInterface.php]: `ZoneMemberInterface[]`?

Comment (Zales0123) [inline src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactoryInterface.php]: Not exactly, the concept is to pass array of members codes, that are used to create zone members and add them to ``Zone``... so probably it should be ``@param array $memberCodes`` :smile:

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 4]
```diff
@@ -98,5 +98,6 @@ public function storeHasAProductPricedAt($productName, $price)
     public function productBelongsToTaxCategory(ProductInterface $product, TaxCategoryInterface $taxCategory)
     {
         $product->setTaxCategory($taxCategory);
+        $this->productRepository->add($product);
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: Actually, we should use a manager here and call flush(). OR ideally flush once after whole background but we can't... :S

File: features/shop/cart_with_multiple_tax_rates.feature [hunk 5]
```diff
@@ -0,0 +1,31 @@
+@ui-cart
+Feature: Apply correct tax for items with different tax rates
+    In order to buy goods with correct taxes applied
```
Comment (michalmarcinkowski) [inline features/shop/cart_with_multiple_tax_rates.feature]: I will try again to make it better `In order to pay proper amount when buying goods from different tax categories`, WDYT?

File: features/shop/cart_with_multiple_tax_rates.feature [hunk 6]
```diff
@@ -0,0 +1,31 @@
+@ui-cart
+Feature: Apply correct tax for items with different tax rates
```
Comment (michalmarcinkowski) [inline features/shop/cart_with_multiple_tax_rates.feature]: IMO the file should be named the same as this line

File: features/shop/cart_with_multiple_tax_rates.feature [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    I want to have correct taxes applied to my order
+
+    Background:
+        Given the store is operating on a single "France" channel
+        And there is "EU" zone containing all members of European Union
+        And default currency is "EUR"
+        And default tax zone is "EU"
+        And store has "EU VAT" tax rate of 23% for "Clothes" within "EU" zone
+        And store has "Low VAT" tax rate of 5% for "Mugs" within "EU" zone
+        And there is user "john@example.com" identified by "password123"
```
Comment (michalmarcinkowski) [inline features/shop/cart_with_multiple_tax_rates.feature]: I would move it at the end of background, just before `I am logged...`

File: features/shop/cart_with_multiple_tax_rates.feature [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        And store has a product "PHP T-Shirt" priced at "€100.00"
+        And product "PHP T-Shirt" belongs to "Clothes" tax category
+        And store has a product "Symfony Mug" priced at "€50.00"
+        And product "Symfony Mug" belongs to "Mugs" tax category
+        And I am logged in as "john@example.com"
+
+    Scenario: Proper taxes for different taxed products
+        When I add product "PHP T-Shirt" to the cart
+        When I add product "Symfony Mug" to the cart
```
Comment (michalmarcinkowski) [inline features/shop/cart_with_multiple_tax_rates.feature]: And

File: src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php [hunk 9]
```diff
@@ -62,11 +62,11 @@ public function createTyped($type)
     /**
      * {@inheritdoc}
      */
-    public function createWithMembers(array $members)
+    public function createWithMembers(array $membersCodes)
     {
         /* @var ZoneInterface $zone */
         $zone = $this->createNew();
-        foreach ($members as $member) {
+        foreach ($membersCodes as $member) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php]: ... as $memberCode

File: src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php [hunk 10]
```diff
@@ -62,11 +62,11 @@ public function createTyped($type)
     /**
      * {@inheritdoc}
      */
-    public function createWithMembers(array $members)
+    public function createWithMembers(array $membersCodes)
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php]: `$memberCodes`?

Comment (pamil) [inline src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php]: It suggests that there is an array of codes belonging to one member, `$membersCodes` is all right.

Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php]: You're correct :)

File: features/shop/cart_with_multiple_tax_rates.feature (hunk 5 above)
Comment (Zales0123) [inline features/shop/cart_with_multiple_tax_rates.feature]: :+1:

File: features/shop/cart_with_multiple_tax_rates.feature (hunk 6 above)
Comment (Zales0123) [inline features/shop/cart_with_multiple_tax_rates.feature]: I was thinking about it, but ``apply_correct_tax_for_items_with_different_tax_rates.feature`` is pretty long filename :smile: We should definitely think about filename <-> feature name convention, as in previously implemented scenarios we have this relation kept (e.g. ``cart_without_taxes.feature`` <-> ``Feature: Cart without taxes``).
The question is, shall we keep this convention even if filename would be long as boa snake. @michalmarcinkowski WDYT?
And if yes, shall previous scenarios be changed, to be more consistent with feature naming (``Feature: Apply correct tax for items with the same tax rate``)?

File: features/shop/cart_with_multiple_tax_rates.feature (hunk 5 above)
Comment (pjedrzejewski) [inline features/shop/cart_with_multiple_tax_rates.feature]: :+1:

File: features/shop/cart_with_multiple_tax_rates.feature (hunk 6 above)
Comment (pjedrzejewski) [inline features/shop/cart_with_multiple_tax_rates.feature]: I am fine with long file names, less vague = better. Feature files should be as decoupled as possible, with clear names and scenarios related to a single feature.

I'd rename ``cart_without_taxes`` to ``not_applying_taxes_for_products_without_tax_category.feature`` or something like that. Cart without taxes is not a feature. :D

File: src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php (hunk 9 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AddressingBundle/Factory/ZoneFactory.php]: Or just ``$code`` as we do not use any other code in this method.

File: features/shop/cart_with_multiple_tax_rates.feature (hunk 6 above)
Comment (Zales0123) [inline features/shop/cart_with_multiple_tax_rates.feature]: Convinced and agreed :+1:

**Comment by pjedrzejewski** [comment]
Thank you Mateusz!

**Pull request merged**